File: utils/py_utils/file_io.py
# Copyright (c) 2025 D-Robotics Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# flake8: noqa: E501


"""
file_io: File and resource I/O utilities.

This module defines a collection of helper functions for interacting with
files and external resources. It serves as a common utility layer for
handling file existence checks, resource loading, and basic data access,
and is intended to be reused across different samples and runtimes.

Key Features:
    - Provide unified interfaces for file and resource access.
    - Encapsulate common file-related operations to reduce duplication.
    - Remain independent of specific models or tasks.

Typical Usage:
    >>> from file_io import download_model_if_needed
    >>> download_model_if_needed(model_path, download_url)

Notes:
    - This module focuses on general-purpose I/O utilities and should not
      include algorithm-specific or business-specific logic.
    - New file or resource related helpers should follow the same design
      principles and be added to this module when appropriate.
"""


import os
import cv2
import numpy as np
from typing import Dict

def download_model_if_needed(model_path: str, download_url: str) -> None:
    """Ensure that the model file exists locally, downloading it if necessary.

    If the model file specified by `model_path` does not exist, this function
    downloads it from the given URL and saves it to the target location.

    Args:
        model_path: Full local path where the model file should be saved.
        download_url: URL used to download the model file.

    Returns:
        None

    Raises:
        RuntimeError: If the download fails or the model file does not exist
            after the download attempt.
    """
    # Return immediately if the model file already exists
    if os.path.exists(model_path):
        print(f"Model already exists: {model_path}")
        return

    # Extract directory and filename from the target model path
    dirname = os.path.dirname(model_path)
    filename = os.path.basename(model_path)

    print(f"File {model_path} does not exist. Downloading model '{filename}'...")

    # Create the directory if it does not already exist
    os.makedirs(dirname, exist_ok=True)

    print(f"Downloading from: {download_url}")
    print(f"Saving to: {model_path}")

    # Execute the wget command to download the model file
    cmd = f"wget -c {download_url} -O {model_path}"
    ret = os.system(cmd)

    # Validate whether the download was successful
    if ret != 0 or not os.path.exists(model_path):
        raise RuntimeError(f"Failed to download model from {download_url}")


def load_image(img_path: str) -> np.ndarray:
    """Load an image from a file path using OpenCV.

    The image is loaded in BGR color format, which is the default format
    used by OpenCV.

    Args:
        img_path: Path to the image file.

    Returns:
        The loaded image as a NumPy array in BGR format.

    Raises:
        FileNotFoundError: If the image cannot be loaded from the given path.
    """
    img = cv2.imread(img_path)
    if img is None:
        raise FileNotFoundError(f"Image file '{img_path}' not found.")
    return img


def load_class_names(path: str) -> list:
    """Load class names from a label file.

    Each line in the file is treated as a single class name. Empty lines
    are ignored.

    Args:
        path: Path to the label file. Each non-empty line contains one
            class name.

    Returns:
        A list of class name strings.
    """
    with open(path, 'r') as f:
        # Strip whitespace and filter out empty lines
        class_names = [line.strip() for line in f.readlines() if line.strip()]
    return class_names


def load_labels(label_path: str) -> Dict[int, str]:
    """
    @brief Load labels from file. Supports dictionary format string or line-separated list.
    @param label_path Path to the label file.
    @return Dictionary of labels {index: name}.
    """
    labels = {}
    if not os.path.exists(label_path):
        print(f"Warning: Label file not found: {label_path}")
        return labels

    try:
        with open(label_path, 'r', encoding='utf-8') as f:
            content = f.read().strip()
        
        if content.startswith('{'):
            # Handle python dictionary string format (e.g. ImageNet labels)
            labels = eval(content)
        else:
            # Handle standard line-separated format
            lines = [line.strip() for line in content.split('\n') if line.strip()]
            labels = {i: name for i, name in enumerate(lines)}
            
    except Exception as e:
        print(f"Warning: Failed to load labels from {label_path}: {e}")

    return labels

File: utils/c_utils/src/file_io.cc
/*
 * Copyright (c) 2025, XiangshunZhao D-Robotics.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file file_io.cc
 * @brief Provide common file input utilities for loading images, labels, and related resources.
 *
 * This file implements a set of generic file I/O helpers used by samples and runtime code.
 * It focuses on reading external data (such as images and label files) into in-memory
 * representations required by the inference pipeline.
 */

#include <fstream>
#include <iostream>
#include <iterator>
#include <regex>
#include <stdexcept>
#include <string>
#include <opencv2/imgcodecs.hpp>

#include "file_io.hpp"

/**
 * @brief Load an image from disk in BGR color space.
 *
 * @param[in] image_file Absolute or relative path to the image file.
 * @return cv::Mat Loaded image in BGR (empty if loading fails).
 */
cv::Mat load_bgr_image(const std::string& image_file) {
    // Read as 3-channel BGR image
    cv::Mat bgr_mat = cv::imread(image_file, cv::IMREAD_COLOR);

    // Warn if failed
    if (bgr_mat.empty()) {
        std::cerr << "ERROR: Failed to load image: " << image_file << std::endl;
    }
    return bgr_mat;
}

/**
 * @brief Load an ImageNet-1000 label map dumped as "{123: 'label'}" text.
 *
 * Expected file format (single object): {0: 'label0', 1: 'label1', ...}
 *
 * @param[in] label_path Path to label file.
 * @return std::map<int,std::string> Mapping from class id to label string. Returns empty map on failure.
 */
std::map<int, std::string> load_imagenet1000_label_map(const std::string &label_path) {
  std::map<int, std::string> label_map;

  std::ifstream infile(label_path);
  if (!infile.is_open()) {
    std::cerr << "Failed to open label file: " << label_path << std::endl;
    return label_map;  // empty
  }

  // Read entire file into a string
  std::string content((std::istreambuf_iterator<char>(infile)),
                      std::istreambuf_iterator<char>());
  infile.close();

  // Trim possible leading '{' and trailing '}'
  if (!content.empty() && content.front() == '{') content.erase(0, 1);
  if (!content.empty() && content.back() == '}') content.pop_back();

  // Regex to match: "<spaces><id><spaces>:<spaces>'<label>'"
  std::regex entry_regex(R"(\s*(\d+)\s*:\s*'([^']*)')");
  auto begin = std::sregex_iterator(content.begin(), content.end(), entry_regex);
  auto end   = std::sregex_iterator();

  // Parse all matches
  for (auto it = begin; it != end; ++it) {
    int id = std::stoi((*it)[1].str());     // group 1: numeric id
    std::string label = (*it)[2].str();     // group 2: label text
    label_map[id] = label;
  }

  return label_map;
}

/**
 * @brief Load labels line-by-line from a plain text file.
 *
 * Each non-empty line is considered one label. Trailing '\r' is stripped.
 *
 * @param[in] filename Path to the label text file.
 * @return std::vector<std::string> All labels in order of appearance.
 */
std::vector<std::string> load_linewise_labels(const std::string& filename)
{
    std::vector<std::string> labels;
    std::ifstream infile(filename);
    std::string line;

    if (!infile.is_open()) {
        std::cerr << "Failed to open label file: " << filename << std::endl;
        return labels; // empty
    }

    while (std::getline(infile, line)) {
        if (!line.empty()) {
            // Remove trailing '\r' if present (Windows line endings)
            if (!line.empty() && line.back() == '\r') {
                line.pop_back();
            }
            labels.push_back(line);
        }
    }
    return labels;
}

/**
 * @brief Load a token->id vocabulary from JSON and produce id->token table.
 *
 * The JSON is expected to be an object: {"<token>": <id>, ... }.
 *
 * @param[in] vocab_file Path to JSON vocabulary file.
 * @return std::vector<std::string> Vector where index is token id and value is token string.
 *
 * @throws std::runtime_error If the file cannot be opened or parsed.
 */
std::vector<std::string> load_id2token(const std::string& vocab_file)
{
    std::ifstream ifs(vocab_file);
    if (!ifs.is_open()) {
        throw std::runtime_error("Failed to open vocab file: " + vocab_file);
    }

    json j;
    ifs >> j;  // Parse JSON to object

    // Find maximum id to size the vector
    int max_id = -1;
    for (auto it = j.begin(); it != j.end(); ++it) {
        int id = it.value().get<int>();
        if (id > max_id) max_id = id;
    }

    // Allocate id->token vector
    std::vector<std::string> id2token(max_id + 1);

    // Reverse the mapping: token->id → id->token
    for (auto it = j.begin(); it != j.end(); ++it) {
        const std::string& token = it.key();
        int id = it.value().get<int>();
        if (id >= 0 && id <= max_id) {
            id2token[id] = token;
        }
    }

    return id2token;
}
